DOCS-2881: Revamp Micro-RDK docs (#4241)
viamrobotics/docs: docs/operate/get-started/setup-micro/_index.md

---
linkTitle: "Set up an ESP32"
title: "Set up an ESP32"
weight: 25
layout: "docs"
type: "docs"
images: ["/installation/thumbnails/install.png"]
imageAlt: "Install Viam"
no_list: false
description: "Install the lightweight version of the software that drives hardware and connects your device to the cloud."
prev: "/operate/get-started/basics/"
next: "/operate/get-started/other-hardware/micro-module/"
aliases:
  - /installation/microcontrollers/
  - /installation/prepare/microcontrollers/
  - /installation/prepare/microcontrollers/
  - /build/micro-rdk/
  - /get-started/installation/microcontrollers/
  - /installation/viam-micro-server-setup/
  - /operate/reference/viam-micro-server/viam-micro-server-troubleshooting/
---

Viam maintains a [lightweight version of `viam-server`](/operate/reference/viam-micro-server/) for microcontrollers.

## Supported microcontrollers

> ESP32 microcontrollers must have at least 2 cores, 384kB SRAM, 2MB PSRAM and 8MB flash to work with Viam.

Reviewer: ```suggestion
To work with Viam's micro-RDK, ESP32 microcontrollers must have at least:

- 2 cores
- 384kB SRAM
- 2MB PSRAM
- 8MB flash
```

[s] I think there are enough items here to warrant an unordered list to make things more scannable. Additionally, an ordered list removes potential reader confusion around the scoping of the 'at least' modifier.

[opt] I think this sentence buries the 'to work with Viam' lede a bit too deep. Otherwise, it's hard to know the point of this sentence when you first start reading it!

Author: Fair! We originally smushed the list into a single line to keep the quickstart further above the fold, but I think it's a tossup which is more important here

The following microcontrollers have been tested with Viam:

- [ESP32-WROVER Series](https://www.espressif.com/en/products/modules/esp32)

## About ESP32 microcontroller setup

> Because microcontrollers do not run operating systems and are instead flashed with firmware, the setup process is different than for regular computers and SBCs.

Reviewer: ```suggestion
SBCs, laptops, and desktops run a complete operating system.
Microcontrollers run dedicated firmware that runs independently of an operating system.
Because microcontrollers do not run an operating system, they use a different setup process than SBCs, laptops, and desktops.
```

[opt] This sentence is doing a lot of work. I took a stab at breaking up the 'operating system vs no OS', 'SBCs/computers vs. microcontrollers', and 'different setup process' ideas up into separate, more readable sentences. I don't love front-loading a sentence that isn't about microcontrollers, but maybe this will give you a better idea.

Author: Smaller change, slightly shorter and keeps the emphasis on microcontrollers. Thoughts?
```suggestion
Microcontrollers do not run operating systems and are instead flashed with dedicated firmware.
Because of this, the microcontroller setup process is different than for SBCs, laptops, and desktops.
```

Reviewer: ship it! looks even better than my suggestion.

   When you update your machine's configuration, `viam-micro-server` automatically gets the updates.

   You are ready to [configure](/operate/get-started/supported-hardware/) any of the components listed above on your machine.

## Build and flash custom firmware

If you are using your ESP32 with resources that are not supported by the pre-built binary, you can build your own firmware with the Micro-RDK and your choice of {{< glossary_tooltip term_id="module" text="modules" >}}.

### Set up your development environment

The [Micro-RDK](https://github.com/viamrobotics/micro-rdk) (from which `viam-micro-server` is built) is written in Rust.
To be able to develop for the Micro-RDK on macOS and Linux systems, you must install the following software on your development machine:

1. Install dependencies:

   {{< tabs >}}
   {{% tab name="Linux" %}}

   ```sh { class="command-line" data-prompt="$"}
   sudo apt-get install bison ccache cmake curl dfu-util flex git gperf libffi-dev libssl-dev libudev-dev libusb-1.0-0 ninja-build python3 python3-pip python3-venv wget
   ```

   {{% /tab %}}
   {{% tab name="macOS" %}}

>    If you haven't already, install Homebrew:
> 
>    ```sh { class="command-line" data-prompt="$"}
>    /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
>    ```

Reviewer: ```suggestion
   If you haven't already, [install Homebrew](https://brew.sh/):
```
[opt] Homebrew's installation instructions, which basically boil down to 'download and immediately run a bash script and give it admin privileges quickly thereafter' has always sketched me out a little. If they ever change, updating our own 'install homebrew' instructions will be kind of annoying. Because of this, I prefer to point users to the Homebrew site to follow the instructions from there, instead.

Author: Okay sure! Removing colon
```suggestion
If you haven't already, [install Homebrew](https://brew.sh/).
```

The above command redirects `espup`'s production of the setup script to `/dev/null` to avoid cluttering the home directory.
If you would like to instead retain the setup script, replace `/dev/null` in the above command with the location where you would like the script to be written, or remove the `-f /dev/null` entirely and the file will be written to `$HOME/export-esp.sh` by default.
{{% /hiddencontent %}}

### Build your firmware

Create firmware that integrates an existing module with the Micro-RDK:

1. Create a new machine and obtain its credentials:

   Navigate to the [Viam app](https://app.viam.com) and add a new machine.
   Click on the name of the machine to go to the machine's page, then select the **CONFIGURE** tab.

   Then select the part status dropdown to the right of your machine's name on the top of the page and copy the **Machine cloud credentials**:

   {{<imgproc src="/get-started/micro-credentials.png" resize="450x" declaredimensions=true alt="Machine part info menu accessed by Live status indicator, with machine cloud credentials button highlighted." class="shadow" >}}

>    The Micro-RDK needs these credentials, which contains your machine part secret key and cloud app address, to connect to the [Viam app](https://app.viam.com).

Reviewer: ```suggestion
   The Micro-RDK needs these credentials, which contain your machine part secret key and cloud app address, to connect to the [Viam app](https://app.viam.com).
```
[s] subject/verb agreement, I think?

Author: oops yep! Does [s] mean spelling?

1. Generate a new project skeleton from [this template](https://github.com/viamrobotics/micro-rdk/tree/main/templates/project):

   ```sh { class="command-line" data-prompt="$"}
   cargo generate --git https://github.com/viamrobotics/micro-rdk.git
   ```

   Follow the prompts:

   - Which sub-template should be expanded?: `templates/project`.
   - Project name: Your choice.
   - MCU: `esp32`.
   - Include camera module?: If you will use an `esp32-camera` or a `fake` camera as a component of your machine, select `true`.

>    - Machine cloud credentials: Paste in the machine cloud credentials you obtained in step 1

Reviewer: ```suggestion
   - Machine cloud credentials: Paste in the machine cloud credentials you obtained in step 1.
```
[s] punctuation

   - Enter WiFi name: The name of the WiFi network for the ESP32 to connect to on boot. Note that you must provide a 2.4 GHz network.

>    - Enter the WiFi password: The password for the 2.GHz WiFi network.

Author: huehue "wifi" in some casing appears over 100 times in our .md files at present compared to 5 times with the hyphen, though we do have a suggestion-level linter thing telling us to hyphenate it. I will change it here, but it's a drop in the bucket until someone feels like making a big nit fix pr about it :D

   The CLI automatically initializes a git repository in the generated directory for version control and in case you want to use cloud build later.

1. Navigate into the generated project:

   ```sh { class="command-line" data-prompt="$"}
   cd <path-to/your-project-directory>
   ```

1. Add any desired modules to the project by including them in the `dependencies` section of the `Cargo.toml` for the generated project:

   {{< tabs >}}
   {{% tab name="Local path dependency" %}}

   While developing a module, you'll typically use a local path dependency to reference your module directory, for example:

   ```toml {class="line-numbers linkable-line-numbers" data-line="3"}
   [dependencies]
   ...
   my-module = { path = "../my-module" }
   ```

   {{% /tab %}}
   {{% tab name="Git repository dependency" %}}

   For modules hosted in a Git repository, specify the repository URL and optionally a specific commit (`rev`), branch (`branch`), or tag (`tag`), for example:

   ```toml {class="line-numbers linkable-line-numbers" data-line="3"}
   [dependencies]
   ...
   my-module = { git = "https://github.com/username/my-module.git", tag = "v1.0.0" }
   ```

>    To use any of the [example modules](https://github.com/viamrobotics/micro-rdk/blob/main/examples/modular-drivers/README.md#example-modules) provided by Viam, use this line, specifying the correct Micro-RDK version that matches your development environment:
> 

Reviewer: ```suggestion
   To use any of the [example modules](https://github.com/viamrobotics/micro-rdk/blob/main/examples/modular-drivers/README.md#example-modules) provided by Viam, use this line, specifying the Micro-RDK version that matches your development environment:

```
[s] correct doesn't add anything here

   ```toml {class="line-numbers linkable-line-numbers" data-line="3"}
   [dependencies]
   ...
   micro-rdk-modular-driver-example = { git = "https://github.com/viamrobotics/micro-rdk.git", rev = "v0.5.0", package = "micro-rdk-modular-driver-example", features = ["esp32"] }
   ```

   {{% /tab %}}
   {{< /tabs >}}

1. Compile the project:

   ```sh { class="command-line" data-prompt="$"}
   make build-esp32-bin
   ```

>    The first build may be fairly time consuming, as ESP-IDF must be cloned and built, and all dependent Rust crates must be fetched and built.
>    Subsequent builds will be faster.

Reviewer: ```suggestion
   The first build can take a few minutes, as ESP-IDF must be cloned and built, and all dependent Rust crates must be fetched and built.
   Subsequent builds will be faster.
```
[s] provide a concrete timeframe, even if it's on the high side, since 'fairly time consuming' could take hours in the hardware world and I don't think this could take hours, even on dial-up!

Author: Subsequent builds also take several minutes :D I like the idea of being more concrete but not sure what the upper bound might look like-- @gvaradarajan ?

### Flash your ESP32

Upload the generated firmware to your ESP32:

1. Use a data cable to connect your ESP32 board to a USB port on your development machine.

1. Run the following command to flash the firmware to your ESP32:

   ```sh { class="command-line" data-prompt="$"}
   make flash-esp32-bin
   ```

   When prompted, select the serial port that your ESP32 is connected to through a data cable.

   If the flash is successful, you will retain a serial connection to the board until you press `Ctrl-C`.
   While the serial connection is live, you can also restart the currently flashed image with `Ctrl-R`.

1. Navigate to your new machine's page on the [Viam app](https://app.viam.com).

>    If successful, the status indicator should turn green and show **Live**.

Reviewer: ```suggestion
   If successful, the status indicator should turn green and show the **Live** status.
```
[opt] provide slightly more context, not a big deal though

### Configure and test your machine

You can now configure the models you included in your firmware and test them:

1. Navigate to your machine's page in the [Viam app](https://app.viam.com).

1. From the **CONFIGURE** tab, click **JSON** mode.
   Micro-RDK components and services must be configured in JSON.

1. Add your components and services to the machine configuration.
   If you are using one of the example modules, find configuration details in the [example modules README](https://github.com/viamrobotics/micro-rdk/blob/main/examples/modular-drivers/README.md#example-modules).
   For example, if you want to use a free heap sensor, your configuration could look like this:

   ```json
   {
     "components": [
       {
         "name": "my-free-heap-sensor",
         "api": "rdk:component:sensor",
         "model": "free-heap",
         "attributes": {}
       }
     ]
   }
   ```

1. Click **Builder** mode and find the configuration card of the component you want to test.

1. Click to open the **Test** section of the card and use the interface to test the component.

> 1. If you make changes to your modules, rerun the build and flash steps to update the firmware and test the new version on your ESP32.

Reviewer: ```suggestion
1. If you make changes to your module implementation, rerun the build and flash steps to update the firmware and test the new version on your ESP32.
```
[s] unless we clarify, I suspect some users might misunderstand and think that they need to re-flash even if they only changed the configuration (or maybe I'm misunderstanding!)

Author: How about this?
```suggestion
1. If you make changes to your module code, rerun the build and flash steps to update the firmware and test the new version on your ESP32.
```

## Cloud build and over-the-air updates

You can use the Viam cloud build service to build and host your firmware, and use the OTA service to remotely update the firmware on your microcontrollers using the cloud-hosted firmware.

### Build firmware with cloud build

When developing and testing your firmware, it is easiest to build firmware locally.
When you are ready to deploy your firmware to a fleet of microcontrollers, you can use the Viam cloud build service to simplify the process of building and hosting your firmware by eliminating the need to build locally and then upload the firmware image to a cloud storage bucket.

To build firmware with cloud build, follow these steps:

1. Create a GitHub repository for any [module you created](/operate/get-started/other-hardware/micro-module/) and want to use in your firmware.
   Upload the contents of your local module directory to the repository.

1. Create a firmware project using the `templates/project` template.

>    _Do not include Wifi credentials or machine cloud credentials in the project_, since you will be publishing the project to a public GitHub repository.

Reviewer: ```suggestion
   _Do not include Wi-Fi credentials or machine cloud credentials in the project_, since you will be publishing the project to a public GitHub repository.
```
[s] Wi-Fi proper name

   The OTA service does not modify the contents of the NVS partition, which contains the machine cloud credentials and WiFi credentials, so OTA updates do not require the credentials.

1. Update your firmware project's `Cargo.toml` file to use the URL of the GitHub repository, for example:

   ```toml
   [dependencies]
   ...
   my-module = { git = "https://github.com/username/my-module.git" }
   ```

1. Create a public GitHub repository for your firmware project, and upload the contents of the firmware project directory to the repository.
   The `templates/project` template you used to create your project contains a GitHub workflow file that is used by the Viam cloud build service to build your firmware.

> 1. In the firmware GitHub repository, navigate to **Settings** &rarr; **Actions** &rarr; **General** and enable **Read and write permissions**.

Author: There's a suggestion but no changes in it? Did you mean to make a change suggestion?

1. Paste in the template below, then configure the URL from which to fetch new firmware, and a version name of your choice.
   The value of the `version` field is not directly used by the OTA service, so you can use any string.

{{< tabs >}}
{{% tab name="JSON Template" %}}

```json {class="line-numbers linkable-line-numbers" data-line="3-10"}
{
  "services": [
    {
      "name": "OTA",
      "api": "rdk:service:generic",
      "model": "rdk:builtin:ota_service",
      "attributes": {
        "url": "<URL where firmware is stored in cloud storage>",
        "version": "<version name>"
      }
    }
  ]
}
```

{{% /tab %}}
{{% tab name="JSON Example" %}}

```json {class="line-numbers linkable-line-numbers"}
{
  "services": [
    {
      "name": "OTA",
      "api": "rdk:service:generic",
      "model": "rdk:builtin:ota_service",
      "attributes": {
        "url": "https://github.com/Jessamy/modulefirmware/releases/download/v0.1.2/modulefirmware-ota.bin",
        "version": "myVersion1"
      }
    }
  ]
}
```

{{% /tab %}}
{{< /tabs >}}

Your device checks for configuration updates periodically.

> If the device receives a configuration with a modified `version` field, the device downloads the new firmware to an inactive partition and restarts.

Author: I'm not sure I know what you mean?

   The Micro-RDK needs these credentials, which contains your machine part secret key and cloud app address, to connect to the [Viam app](https://app.viam.com).

   {{% snippet "secret-share.md" %}}

1. Generate a new project skeleton from [this template](https://github.com/viamrobotics/micro-rdk/tree/main/templates/project):

   ```sh { class="command-line" data-prompt="$"}
   cargo generate --git https://github.com/viamrobotics/micro-rdk.git
   ```

   Follow the prompts:

   - Which sub-template should be expanded?: `templates/project`.
   - Project name: Your choice.
   - MCU: `esp32`.
   - Include camera module?: If you will use an `esp32-camera` or a `fake` camera as a component of your machine, select `true`.
   - Machine cloud credentials: Paste in the machine cloud credentials you obtained in step 1
   - Enter WiFi name: The name of the WiFi network for the ESP32 to connect to on boot. Note that you must provide a 2.4 GHz network.
   - Enter the WiFi password: The password for the 2.GHz WiFi network.

   Hit **Enter** to generate the project.

>    The CLI automatically initializes a git repository in the generated directory for version control and in case you want to use cloud build later.

Author: Or should I even go so far as to remove "for version control" as well?
